packages/notifi-react-card/lib/hooks/useIntercomChat.ts

```diff
@@ -1,32 +1,148 @@
 
 import { useNotifiClientContext } from '../context';
+import { formatConversationDateTimestamp } from '../utils/datetimeUtils';
+
+export type ChatConversation = FeedEntry &
+  Readonly<{
+    feed: FeedEntry[];
+    createdDate: string | undefined;
+    lastMessage: ChatMessage | undefined;
+  }>;
+
+export type ChatMessage = Readonly<{
+  id: string;
+  message: string;
+  userId: string;
+  createdDate: string;
+  updatedDate: string;
+  conversationId: string;
+}>;
+
+export type MessageDirection = 'INCOMING' | 'OUTGOING';
+
+export type MessagesBlockFeedEntry = Readonly<{
+  type: 'MESSAGES_BLOCK';
+  direction: MessageDirection;
+  messages: ChatMessage[];
+}>;
+
+export type FeedEntry = {
+  id: string;
+  timestamp: string;
+} & MessagesBlockFeedEntry;
 
 type GetConversationMessagesInputProps = Readonly<{
   first?: number;
-  after?: string;
   conversationId: string;
 }>;
 
-export const useConversationMessages = ({
-  first,
-  after,
+export const useIntercomChat = ({
+  first = 5,
   conversationId,
 }: GetConversationMessagesInputProps) => {
+  const [chatMessages, setChatMessages] = useState<ConversationMessagesEntry[]>(
+    [],
+  );
+  const [endCursor, setEndCursor] = useState<string | undefined>();
+  const [hasNextPage, setHasNextPage] = useState<boolean | null>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [atTop, setAtTop] = useState<boolean>(false);
+  const [visibleRange, setVisibleRange] = useState<ListRange>({
+    startIndex: 0,
+    endIndex: 0,
+  });
+  const [isScrolling, setIsScrolling] = useState<boolean | null>();
   const { client } = useNotifiClientContext();
 
-  const input = {
-    first: first,
-    after: after,
-    getConversationMessagesInput: { conversationId },
-  };
+  const getConversationMessages = useCallback(() => {
+    setIsLoading(true);
+    client
+      .getConversationMessages({
+        first,
+        after: endCursor,
+        getConversationMessagesInput: { conversationId },
+      })
+      .then((response) => {
+        if (Array.isArray(response.nodes)) {
+          setChatMessages([...chatMessages, ...response.nodes]);
+        }
+        if (response.pageInfo) {
+          setEndCursor(response.pageInfo.endCursor);
+          setHasNextPage(response.pageInfo.hasNextPage);
+        }
+        setIsLoading(false);
+      });
+  }, [conversationId, endCursor, chatMessages]);
 
-  const getConversationMessages = useCallback(async () => {
-    const response = await client.getConversationMessages(input);
-    return response;
+  // initialization - load first batch of messages
+  useEffect(() => {
+    getConversationMessages();
   }, []);
 
+  // load more messages when scrolling to the top
+  useEffect(() => {
+    if (hasNextPage && atTop && isScrolling && visibleRange.startIndex === 0) {
+      getConversationMessages();
+    }
+  }, [hasNextPage, atTop, isScrolling, visibleRange.startIndex]);
+
+  const conversation = useMemo(() => {
```

Comment:
Can you please explain what is happening with `conversation`?

Reply:
So the conversation is the information we need to render the data, like message direction and messages feed, all the logics are adding here for getting the needed data(followed the pattern of xMS chat in hub).

Reply:
Thank you. As we chatted in our huddle, I would suggest that you rename the variable and also put the if statement condition in a more readable `const`. A comment would be helpful as well!